### ISSUE TITLE
Convert libmemcached to an external repository rather than using the system library [phab:D1703]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2132]
_Created by **shahms** at 2017-06-09 16:09:39_: 

Don't rely on the system libmemcached

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2017-06-09 16:53:17_:

```
ERROR: /private/var/tmp/_bazel_salguarnieri/ca16ea5a435d8ac2288f379cdf092007/external/org_libmemcached_libmemcached/BUILD.bazel:95:1: C++ compilation of rule 'org_libmemcached_libmemcached//:libmemcached' failed: Process exited with status 1 [sandboxed].
In file included from external/org_libmemcached_libmemcached/libmemcached/stats.cc:37:
In file included from external/org_libmemcached_libmemcached/libmemcached/common.h:135:
external/org_libmemcached_libmemcached/libmemcached/memcached/protocol_binary.h:39:10: fatal error: 'libmemcachedprotocol-0.0/vbucket.h' file not found
#include <libmemcachedprotocol-0.0/vbucket.h>
```

I swapped in a mem_config.h from a local run of configure and I get the same error.

**Comment by kythe** [comment]
_**shahms** wrote at 2017-06-09 17:11:38_:

>>! In (earlier comment), salguarnieri wrote:
> ```
> ERROR: /private/var/tmp/_bazel_salguarnieri/ca16ea5a435d8ac2288f379cdf092007/external/org_libmemcached_libmemcached/BUILD.bazel:95:1: C++ compilation of rule 'org_libmemcached_libmemcached//:libmemcached' failed: Process exited with status 1 [sandboxed].
> In file included from external/org_libmemcached_libmemcached/libmemcached/stats.cc:37:
> In file included from external/org_libmemcached_libmemcached/libmemcached/common.h:135:
> external/org_libmemcached_libmemcached/libmemcached/memcached/protocol_binary.h:39:10: fatal error: 'libmemcachedprotocol-0.0/vbucket.h' file not found
> #include <libmemcachedprotocol-0.0/vbucket.h>
> ```
> 
> I swapped in a mem_config.h from a local run of configure and I get the same error.

Ugh, thanks.  I think bazel was picking up my locally installed copy for this missing header.  Uninstalling the package revealed a few other issues.  Re-running the tests now and will update.

**Comment by kythe** [comment]
_**shahms** wrote at 2017-06-09 17:18:12_:

  - Include libmemcachedprotocol dependency and output to genfiles so header can be found

**Comment by kythe** [comment]
_**shahms** wrote at 2017-06-09 17:18:36_:

Sal, PTAL?

**Comment by kythe** [comment]
_**shahms** wrote at 2017-06-09 17:45:16_:

  - Update mem_config.h and BUILD to enable appropriate features on macOS and Linux

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-06-09 18:03:11_:

Sometimes for comments like this we do: /* undef ... */, sometimes we define to 0. I'm pretty sure this comes from autoconf, but if it doesn't can you explain why we define to 0 sometimes?

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2017-06-09 18:03:11_:

Can you add a comment that this file was locally modified to work for linux and macOS?

**Comment by kythe** [comment]
_**salguarnieri** accepted this revision at 2017-06-09 18:03:11_

**Comment by kythe** [comment]
_**shahms** wrote at 2017-06-09 18:06:54_:

Done on the comment.  The second is "because autoconf" though (the macros themselves are inconsistently used, sometimes `#ifdef`, sometimes `if(MACRO_VALUE)`.)

**Comment by kythe** [comment]
_**shahms** wrote at 2017-06-09 18:12:57_:

  - Add comment to mem_config.h about manual modifications

**Comment by kythe** [comment]
_**shahms** closed via commit 365a10b518251423e71f27caf7cb276ed08a9feb at 2017-06-09 18:15:29_
